File: ports/php/src/Security/Crypto/Bip39Canonical.php
<?php

declare(strict_types=1);

namespace UnicodePhp\Security\Crypto;

use UnicodePhp\Data;
use UnicodePhp\Security\Identity\Locale;
use UnicodePhp\Security\Identity\Ucd;
use UnicodePhp\Utf8;

final class Bip39Verdict
{
    /** @param list<int> $positions @param list<int> $canonical */
    public function __construct(
        public readonly ?string $sub,
        public readonly array $positions,
        public readonly ?string $language,
        public readonly array $canonical,
        public readonly int $wordCount,
    ) {
    }
}

final class Bip39Canonical
{
    private const WORDLIST_FILES = [
        ['english', 'english.txt'],
        ['japanese', 'japanese.txt'],
        ['korean', 'korean.txt'],
        ['spanish', 'spanish.txt'],
        ['chinese_simplified', 'chinese_simplified.txt'],
        ['chinese_traditional', 'chinese_traditional.txt'],
        ['french', 'french.txt'],
        ['italian', 'italian.txt'],
        ['czech', 'czech.txt'],
        ['portuguese', 'portuguese.txt'],
    ];

    /** @var list<array{name:string,set:array<string,bool>}>|null */
    private static ?array $wordlists = null;

    /** @param list<int> $cps */
    private static function key(array $cps): string
    {
        return implode(',', $cps);
    }

    private static function wordlistSet(string $raw): array
    {
        $set = [];
        foreach (explode("\n", $raw) as $line) {
            $line = rtrim($line, "\r");
            if ($line !== '') {
                $set[self::key(Utf8::decodeToCodepoints(array_values(unpack('C*', $line) ?: [])))] = true;
            }
        }
        return $set;
    }

    /** @return list<array{name:string,set:array<string,bool>}> */
    private static function wordlists(): array
    {
        if (self::$wordlists !== null) {
            return self::$wordlists;
        }
        $out = [];
        foreach (self::WORDLIST_FILES as [$name, $file]) {
            $out[] = ['name' => $name, 'set' => self::wordlistSet(Data::read('bip39/' . $file))];
        }
        self::$wordlists = $out;
        return $out;
    }

    /** @param list<int> $canonical @return list<list<int>> */
    private static function splitWords(array $canonical): array
    {
        $words = [];
        $current = [];
        foreach ($canonical as $cp) {
            if ($cp === 0x20) {
                if ($current !== []) {
                    $words[] = $current;
                    $current = [];
                }
            } else {
                $current[] = $cp;
            }
        }
        if ($current !== []) {
            $words[] = $current;
        }
        return $words;
    }

    /** @param list<int> $word @return list<string> */
    private static function wordlistsContaining(array $word): array
    {
        $out = [];
        $key = self::key($word);
        foreach (self::wordlists() as $wl) {
            if (isset($wl['set'][$key])) {
                $out[] = $wl['name'];
            }
        }
        return $out;
    }

    /** @param list<list<int>> $words */
    private static function uniqueLanguage(array $words): ?string
    {
        foreach (self::wordlists() as $wl) {
            $ok = true;
            foreach ($words as $word) {
                if (!isset($wl['set'][self::key($word)])) {
                    $ok = false;
                    break;
                }
            }
            if ($ok) {
                return $wl['name'];
            }
        }
        return null;
    }

    private static function bip39Whitespace(int $cp): bool
    {
        return $cp === 0x20 || $cp === 0x3000;
    }

    /** @param list<int> $cps @return list<int> */
    private static function collapseWhitespaceToSingle(array $cps): array
    {
        $out = [];
        $inWs = false;
        foreach ($cps as $cp) {
            if (self::bip39Whitespace($cp)) {
                if (!$inWs) {
                    $out[] = 0x20;
                }
                $inWs = true;
            } else {
                $out[] = $cp;
                $inWs = false;
            }
        }
        return $out;
    }

    /** @param list<int> $cps @return list<int> */
    private static function trimLeadingTrailing(array $cps): array
    {
        $start = null;
        foreach ($cps as $i => $cp) {
            if ($cp !== 0x20) {
                $start = $i;
                break;
            }
        }
        if ($start === null) {
            return [];
        }
        $end = $start;
        for ($i = count($cps) - 1; $i >= $start; $i--) {
            if ($cps[$i] !== 0x20) {
                $end = $i;
                break;
            }
        }
        return array_slice($cps, $start, $end - $start + 1);
    }

    /** @param list<int> $cps @return list<int> */
    public static function bip39Canonical(array $cps): array
    {
        return self::trimLeadingTrailing(self::collapseWhitespaceToSingle(Ucd::toLower(Locale::Default, Ucd::toNfkd($cps))));
    }

    /** @param list<int> $cps */
    private static function countTrailingWhitespace(array $cps): int
    {
        $count = 0;
        for ($i = count($cps) - 1; $i >= 0; $i--) {
            if (!self::bip39Whitespace($cps[$i])) {
                break;
            }
            $count++;
        }
        return $count;
    }

    /** @param list<int> $cps */
    private static function firstUppercasePos(array $cps): ?int
    {
        foreach ($cps as $i => $cp) {
            if ($cp >= 0x41 && $cp <= 0x5A) {
                return $i;
            }
        }
        return null;
    }

    /** @param list<int> $cps */
    private static function firstWhitespaceRunPos(array $cps): ?int
    {
        $count = count($cps);
        for ($i = 0; $i < $count; $i++) {
            if (self::bip39Whitespace($cps[$i]) && ($i === 0 || ($i < $count - 1 && self::bip39Whitespace($cps[$i + 1])))) {
                return $i;
            }
        }
        return null;
    }

    /** @param list<int> $a @param list<int> $b */
    private static function firstArrayDivergence(array $a, array $b): ?int
    {
        $n = min(count($a), count($b));
        for ($i = 0; $i < $n; $i++) {
            if ($a[$i] !== $b[$i]) {
                return $i;
            }
        }
        return count($a) === count($b) ? null : $n;
    }

    /** @param list<int> $input */
    public static function detect(array $input): Bip39Verdict
    {
        $canonical = self::bip39Canonical($input);
        $words = self::splitWords($canonical);
        $wordCount = count($words);
        $trailingCount = self::countTrailingWhitespace($input);
        $uppercasePos = self::firstUppercasePos($input);
        $whitespacePos = self::firstWhitespaceRunPos($input);
        $nfkd = Ucd::toNfkd($input);
        $nonNfkdPos = $input === $nfkd ? null : self::firstArrayDivergence($input, $nfkd);
        $unknownIdx = null;
        foreach ($words as $i => $word) {
            if (self::wordlistsContaining($word) === []) {
                $unknownIdx = $i;
                break;
            }
        }

        if ($trailingCount > 0) {
            return new Bip39Verdict('TrailingWhitespace', [count($input) - $trailingCount], null, $canonical, $wordCount);
        }
        if ($uppercasePos !== null) {
            return new Bip39Verdict('MixedCase', [$uppercasePos], null, $canonical, $wordCount);
        }
        if ($whitespacePos !== null) {
            return new Bip39Verdict('WhitespaceAnomaly', [$whitespacePos], null, $canonical, $wordCount);
        }
        if ($nonNfkdPos !== null) {
            return new Bip39Verdict('NonNFKD', [$nonNfkdPos], null, $canonical, $wordCount);
        }
        if ($unknownIdx !== null) {
            return new Bip39Verdict('WordlistMismatch', [$unknownIdx], null, $canonical, $wordCount);
        }
        $lang = self::uniqueLanguage($words);
        if ($lang === null) {
            return new Bip39Verdict('LanguageAmbiguous', [], null, $canonical, $wordCount);
        }
        return new Bip39Verdict(null, [], $lang, $canonical, $wordCount);
    }
}

File: ports/rust/src/security/crypto/bip39_canonical.rs
//! bip39-canonical: BIP-39 mnemonic canonicalisation + wordlist checks.
//!
//! Mirrors `Unicode.Security.Crypto.Bip39Canonical`. Canonical form is
//! NFKD -> to_lower(default) -> collapse BIP-39 whitespace -> trim; `detect`
//! runs six probes in priority order over the input and its canonical words.

use std::collections::HashSet;
use std::sync::OnceLock;

use crate::security::identity::ucd::{self, Locale};

const ENGLISH: &str = include_str!("../../../data/bip39/english.txt");
const JAPANESE: &str = include_str!("../../../data/bip39/japanese.txt");
const KOREAN: &str = include_str!("../../../data/bip39/korean.txt");
const SPANISH: &str = include_str!("../../../data/bip39/spanish.txt");
const CHINESE_SIMPLIFIED: &str = include_str!("../../../data/bip39/chinese_simplified.txt");
const CHINESE_TRADITIONAL: &str = include_str!("../../../data/bip39/chinese_traditional.txt");
const FRENCH: &str = include_str!("../../../data/bip39/french.txt");
const ITALIAN: &str = include_str!("../../../data/bip39/italian.txt");
const CZECH: &str = include_str!("../../../data/bip39/czech.txt");
const PORTUGUESE: &str = include_str!("../../../data/bip39/portuguese.txt");

/// One bip39-canonical scan result.  `sub` is `None` for a clear input (with
/// `language` set to the covering wordlist); otherwise it carries the
/// sub-threat tag and its position(s).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Detection {
    /// Sub-threat tag, or `None` when the mnemonic is canonical and covered.
    pub sub: Option<&'static str>,
    /// Position(s) implicated by the sub-threat.
    pub positions: Vec<usize>,
    /// Covering wordlist language when clear, else `None`.
    pub language: Option<&'static str>,
    /// The BIP-39 canonical form of the input.
    pub canonical: Vec<u32>,
    /// Number of whitespace-separated canonical words.
    pub word_count: usize,
}

fn wordlist_set(raw: &str) -> HashSet<Vec<u32>> {
    raw.lines()
        .filter(|line| !line.is_empty())
        .map(|line| line.chars().map(|c| c as u32).collect())
        .collect()
}

/// The ten wordlists as codepoint-sequence sets, in
/// `Unicode.Generated.BIP39.allLanguages` order (English first).
fn wordlists() -> &'static Vec<(&'static str, HashSet<Vec<u32>>)> {
    static T: OnceLock<Vec<(&'static str, HashSet<Vec<u32>>)>> = OnceLock::new();
    T.get_or_init(|| {
        vec![
            ("english", wordlist_set(ENGLISH)),
            ("japanese", wordlist_set(JAPANESE)),
            ("korean", wordlist_set(KOREAN)),
            ("spanish", wordlist_set(SPANISH)),
            ("chinese_simplified", wordlist_set(CHINESE_SIMPLIFIED)),
            ("chinese_traditional", wordlist_set(CHINESE_TRADITIONAL)),
            ("french", wordlist_set(FRENCH)),
            ("italian", wordlist_set(ITALIAN)),
            ("czech", wordlist_set(CZECH)),
            ("portuguese", wordlist_set(PORTUGUESE)),
        ]
    })
}

fn split_words(canonical: &[u32]) -> Vec<Vec<u32>> {
    let mut words = Vec::new();
    let mut current = Vec::new();
    for &cp in canonical {
        if cp == 0x0020 {
            if !current.is_empty() {
                words.push(std::mem::take(&mut current));
            }
        } else {
            current.push(cp);
        }
    }
    if !current.is_empty() {
        words.push(current);
    }
    words
}

fn wordlists_containing(word: &[u32]) -> Vec<&'static str> {
    wordlists()
        .iter()
        .filter(|(_, set)| set.contains(word))
        .map(|(name, _)| *name)
        .collect()
}

fn unique_language(words: &[Vec<u32>]) -> Option<&'static str> {
    wordlists()
        .iter()
        .find(|(_, set)| words.iter().all(|w| set.contains(w)))
        .map(|(name, _)| *name)
}

fn is_bip39_whitespace(cp: u32) -> bool {
    cp == 0x0020 || cp == 0x3000
}

fn collapse_whitespace_to_single(cps: &[u32]) -> Vec<u32> {
    let mut out = Vec::new();
    let mut in_ws = false;
    for &cp in cps {
        if is_bip39_whitespace(cp) {
            if !in_ws {
                out.push(0x0020);
            }
            in_ws = true;
        } else {
            out.push(cp);
            in_ws = false;
        }
    }
    out
}

fn trim_leading_trailing(cps: &[u32]) -> Vec<u32> {
    let start = match cps.iter().position(|&cp| cp != 0x0020) {
        Some(s) => s,
        None => return Vec::new(),
    };
    let end = cps
        .iter()
        .rposition(|&cp| cp != 0x0020)
        .map_or(start, |i| i + 1);
    cps[start..end].to_vec()
}

fn bip39_canonical(cps: &[u32]) -> Vec<u32> {
    let nfkd = ucd::to_nfkd(cps);
    let lowered = ucd::to_lower(Locale::Default, &nfkd);
    let collapsed = collapse_whitespace_to_single(&lowered);
    trim_leading_trailing(&collapsed)
}

fn count_trailing_whitespace(cps: &[u32]) -> usize {
    cps.iter()
        .rev()
        .take_while(|&&cp| is_bip39_whitespace(cp))
        .count()
}

fn first_uppercase_pos(cps: &[u32]) -> Option<usize> {
    cps.iter().position(|&cp| (0x41..=0x5A).contains(&cp))
}

fn first_whitespace_run_pos(cps: &[u32]) -> Option<usize> {
    for (i, &cp) in cps.iter().enumerate() {
        if is_bip39_whitespace(cp) {
            if i == 0 {
                return Some(i);
            }
            if let Some(&next) = cps.get(i + 1) {
                if is_bip39_whitespace(next) {
                    return Some(i);
                }
            }
        }
    }
    None
}

fn first_array_divergence(a: &[u32], b: &[u32]) -> Option<usize> {
    let n = a.len().min(b.len());
    for (i, (x, y)) in a.iter().zip(b.iter()).enumerate() {
        if x != y {
            return Some(i);
        }
    }
    if a.len() != b.len() {
        Some(n)
    } else {
        None
    }
}

/// Detect a non-canonical or wordlist-mismatched BIP-39 mnemonic.  Six probes
/// in priority order (first hit wins), mirroring `Bip39Canonical.detect`.
pub fn detect(input: &[u32]) -> Detection {
    let canonical = bip39_canonical(input);
    let words = split_words(&canonical);
    let word_count = words.len();

    let trailing_count = count_trailing_whitespace(input);
    let uppercase_pos = first_uppercase_pos(input);
    let whitespace_pos = first_whitespace_run_pos(input);
    let nfkd = ucd::to_nfkd(input);
    let non_nfkd_pos = if input == nfkd.as_slice() {
        None
    } else {
        first_array_divergence(input, &nfkd)
    };
    let first_unknown_idx = words
        .iter()
        .position(|w| wordlists_containing(w).is_empty());

    let (sub, positions, language): (Option<&'static str>, Vec<usize>, Option<&'static str>) =
        if trailing_count > 0 {
            (
                Some("TrailingWhitespace"),
                vec![input.len() - trailing_count],
                None,
            )
        } else if let Some(p) = uppercase_pos {
            (Some("MixedCase"), vec![p], None)
        } else if let Some(p) = whitespace_pos {
            (Some("WhitespaceAnomaly"), vec![p], None)
        } else if let Some(p) = non_nfkd_pos {
            (Some("NonNFKD"), vec![p], None)
        } else if let Some(idx) = first_unknown_idx {
            (Some("WordlistMismatch"), vec![idx], None)
        } else {
            match unique_language(&words) {
                Some(lang) => (None, Vec::new(), Some(lang)),
                None => (Some("LanguageAmbiguous"), Vec::new(), None),
            }
        };

    Detection {
        sub,
        positions,
        language,
        canonical,
        word_count,
    }
}

#[cfg(test)]
mod tests {
    use super::detect;

    const ABANDON: &[u32] = &[0x61, 0x62, 0x61, 0x6E, 0x64, 0x6F, 0x6E];
    const ABOUT: &[u32] = &[0x61, 0x62, 0x6F, 0x75, 0x74];

    fn tag(input: &[u32]) -> Option<&'static str> {
        detect(input).sub
    }

    #[test]
    fn detect_hazard_tags() {
        let mut trailing = ABANDON.to_vec();
        trailing.push(0x20);
        assert_eq!(tag(&trailing), Some("TrailingWhitespace"));
        assert_eq!(
            tag(&[0x41, 0x62, 0x61, 0x6E, 0x64, 0x6F, 0x6E]),
            Some("MixedCase")
        );
        let mut dbl = ABANDON.to_vec();
        dbl.extend_from_slice(&[0x20, 0x20]);
        dbl.extend_from_slice(ABOUT);
        assert_eq!(tag(&dbl), Some("WhitespaceAnomaly"));
        let mut lead = vec![0x20];
        lead.extend_from_slice(ABANDON);
        assert_eq!(tag(&lead), Some("WhitespaceAnomaly"));
        assert_eq!(tag(&[0xFB00]), Some("NonNFKD")); // ﬀ ligature
        assert_eq!(tag(&[0x61, 0x00A0, 0x62]), Some("NonNFKD")); // NBSP
        assert_eq!(tag(&[0x71, 0x7A, 0x71, 0x7A]), Some("WordlistMismatch"));
    }

    #[test]
    fn detect_positions_and_empty_clear() {
        let mut trailing = ABANDON.to_vec();
        trailing.push(0x20);
        assert_eq!(detect(&trailing).positions, vec![7]);
        assert_eq!(
            detect(&[0x41, 0x62, 0x61, 0x6E, 0x64, 0x6F, 0x6E]).positions,
            vec![0]
        );
        let empty = detect(&[]);
        assert_eq!(empty.sub, None);
        assert_eq!(empty.language, Some("english"));
    }

    #[test]
    fn detect_clear_12word_english() {
        let mut mnemonic = Vec::new();
        for _ in 0..11 {
            mnemonic.extend_from_slice(ABANDON);
            mnemonic.push(0x20);
        }
        mnemonic.extend_from_slice(ABOUT);
        let verdict = detect(&mnemonic);
        assert_eq!(verdict.sub, None);
        assert_eq!(verdict.language, Some("english"));
        assert_eq!(verdict.word_count, 12);
    }
}
